Add Result utility class

PiotrSquaresMoscicki opened the pull request:
Based on https://doc.rust-lang.org/std/result/enum.Result.html
I'll add comments to this class soon in this pull request.

PiotrSquaresMoscicki committed:
Add Result draft and tests file

PiotrSquaresMoscicki committed:
Add IsOk tests

PiotrSquaresMoscicki committed:
Add IsErr, GetValue and GetError tests

rybochodonezzar commented:
Is this final form? i would suggest to add several helper functions here:
(assuming `Result`  = `R` because i'm lazy)
`R<T2,E> bind(R<T1,E> x, R<T2,E>(T1) f)` (monadic bind)
`R<T,E> join(R<R<T,E>,E>)` (monadic join)
variadic versions of rust's `and` and `or` (but i would call them `first` and `last`
`template <typename Res, typename E, typename ... Xs> auto lift(Res(Xs ... args) f)`
where `auto` is lambda taking all all arguments lifted (ie.  T replaced with R<T,E>) and R<Res,E> with first error or result (perhaps i should do this one?)
also if `lift` is implemented it could make sense to lift all non-assigning operators

Commit:
Minor methods renames and join impl

Commit:
Add join iomplementation

Commit:
Implement bind

Commit:
add first implementation

Commit:
Rename first to and

Commit:
Add or implementation

Commit:
Remove random debug piece of code

rybochodonezzar commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -10,20 +10,39 @@ namespace pe::core::utils
+		const E& getError() const { return std::get<E>(m_value); }
+
+		template <typename U>
+		Result<U, E> and(Result<U, E> res)
```
Why not make those functions standalone and variadic?

MuniuDev replied:
I suggest simply using `operator&` and `operator|`.

rybochodonezzar replied:
I suggest instead to define for all operators (let's use `+` for example here)
`template <typename X1, typename X2, typename Y, typename E> R<Y,E> operator+(R<X1,E> x1, R<X2,E> x2)`
returning error of `x1` or error of `x2` or `x1 + x2`

then, as mentioned before
`template <typename T, typename E> R<T, E> first(R<T, E> ... xs)`
and overload for ranges
`template <typename T, typename E, template<typename> typename C> R<T,E> first(C<R<T,E>> range)`

rybochodonezzar replied:
or perhaps ditch `first`/`last`, since those functions discard potentially valuable error information unless `E = void`. Perhaps it would be usefull to separate this case into another class

PiotrSquaresMoscicki replied:
resolved - first and last were removed

PiotrSquaresMoscicki replied:
I've removed these functions

rybochodonezzar commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -33,18 +52,67 @@ namespace pe::core::utils
+			return res.getError();
+	}
+
+	using aaa = void(*)();
```
Whaaat is this for?

PiotrSquaresMoscicki replied:
random line i forgot to remove, its already removed

rybochodonezzar commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -33,18 +52,67 @@ namespace pe::core::utils
+	using aaa = void(*)();
+
+	template <typename T1, typename T2, typename E>
+	Result<T2, E> bind(Result<T1, E> res, Result<T2, E> (*func)(T1))
```
This takes only function pointer. You casting closure (lambda with capture) to pointer is UB. Use auto

MuniuDev replied:
I think auto in an function argument declaration is C++20 feature. Use std::function.
Also, this as well might be a simple method.

MuniuDev replied:
Also please change T1 function argument to r-value ref and add std::move  for line 112.

rybochodonezzar commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -10,20 +10,39 @@ namespace pe::core::utils
+		bool isErr() const { return !isOk(); }
+		
+		T& getValue() { return std::get<T>(m_value); }
+		E& getError() { return std::get<E>(m_value); }
```
maybe T& getDefault(T&) (returning default if error) and
T& getHandle(T&(E&)) would be usefull to add?

MuniuDev replied:
I suggest different naming scheme:
`value()`
`error()`
`valueOr(T)`

What `getHandle` is supposed to do?

rybochodonezzar replied:
like `getDefault`/`valueOr` but uses function taking error and returning default value instead.

PiotrSquaresMoscicki replied:
done

Commit:
Add lift

Commit:
Rename or to last and and to first

In cpp standard there are 'and' and 'or' keywords. MSVC seemed to ignore
them but we need to change this for gcc.

Commit:
Use auto in lift function

MuniuDev commented on PolyEngine/CMakeListsCommon.cmake:
```diff
@@ -86,19 +86,19 @@ else()
 find_package(Qt5 COMPONENTS Core Widgets)
 if(NOT DEFINED Qt5_FOUND OR NOT Qt5_FOUND)
-	message(FATAL_ERROR "Qt5 not found. When using a standalone Qt5 toolchain, please setup QTDIR environment variable to point to it.")
+	message(FATAL_ERROR "Qt5 not found. When using a standalone Qt5 toolchain, please setup POLY_QTDIR environment variable to point to it.")
```
This looks like your internal local change. Revert this file.

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,117 @@
+	}
+
+	template <typename E>
+	Result<void, E> join(Result<Result<void, E>, E> res)
```
Are `join` and `bind` widely used names or are we free to name those functions as we want? I would suggest `collapse` instead of `join`. WDYT?
Also, this as well might be a simple method.

JDuchniewicz replied:
FYI there is a std::bind

rybochodonezzar replied:
`join` and `bind` are widely used unless someone is trying to call monads [Warm and fuzzy things](https://www.urbandictionary.com/define.php?term=Warm%20Fuzzy%20Thing) insetead

rybochodonezzar replied:
`std::bind` is completly different thing.
I don't think it should be a simple method. Once `c++20` is a thing, it won't be long before someone somewhere (most probably in boost) defines monad concept. We should use it then

PiotrSquaresMoscicki replied:
we can have namespace result for free functions for result

MuniuDev replied:
What is the use case for this function? I don't see a situation where you can end with a Result with Result inside, as this would mean that you didn't handle the error which you should have handled.

PiotrSquaresMoscicki replied:
resolved - i've removed join

rybochodonezzar replied:
The idea here was to have this as a part of monadic interface, but true, this is practically less usefull than bind

MuniuDev reviewed (changes requested):
I would also like an implicit bool cast to be added.

rybochodonezzar commented:
@MuniuDev agree, but should non-error value be always treated as `true` or should it evaluate to bool value of underlying value? In other words should `(bool)(R(false))` evaluate to `true` or `false`?
I'm on the side of evaluating to `false`, since the other option just duplicates functionality of `isErr()`

PiotrMoscicki commented:
@MuniuDev @rybochodonezzar I wouldn't go for any implicit cast because user will wonder what does it evaluate. Similarly to std::optional which doesn't have implicit bool cast and i've never had a problem with it.

PiotrSquaresMoscicki committed:
Add separators

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,117 @@
+				return Result<void, F>();
+			else
+				return res;
+		}
```
We need to check the behavior of those functions in Rust, as current implementation is strange.

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,117 @@
+	};
+
+	template <typename T, typename E>
+	Result<T, E> join(Result<Result<T, E>, E> res)
```
Please test those functions/methods for number of copies/moves they perform upon execution both on T and on E.

PiotrSquaresMoscicki replied:
done

MuniuDev commented:
Also please add doxygen documentation to all pubic API.

PiotrSquaresMoscicki committed:
Merge branch 'dev' into resultWip

PiotrSquaresMoscicki committed:
Add Ok and Err classes

PiotrSquaresMoscicki committed:
Implemented Result, writing tests in progress

PiotrSquaresMoscicki committed:
Test all member functionalities

PiotrSquaresMoscicki committed:
Remove bind

PiotrSquaresMoscicki committed:
Revert changes in CMakeListsCommon.cmake

PiotrSquaresMoscicki committed:
Add deduction guides to fix g++ compilation

JDuchniewicz commented:
Just a general comment, I suggest better naming for commits, so that we maintain high quality like other repositories :)

https://chris.beams.io/posts/git-commit/

PiotrSquaresMoscicki committed:
WIP Restore bind

PiotrSquaresMoscicki committed:
Restore bind

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,280 @@
+		constexpr bool isErr() const { return !isOk(); }
+		
+		// @returns valid value if has valid value, crashes otherwise :(
+		constexpr T& value() { return std::get<T>(m_value); }
```
Let's get rid of mutable API all together.

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,280 @@
+		constexpr T& value() { return std::get<T>(m_value); }
+
+		// @returns valid value if has valid value, crashes otherwise :(
+		constexpr const T& value() const { return std::get<T>(m_value); }
```
Create change value() to be similar in nature to valueOr() (return T by val, overloaded for l and r-value references).
Add valueRef() which returns const T&.

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,280 @@
+		}
+
+		// returns error if this stores an error, crashes otherwise :(
+		constexpr E& error() { return std::get<E>(m_value); }
```
Do the same as for value().

PiotrSquaresMoscicki committed:
Compilation fixes for result

PiotrSquaresMoscicki committed:
Merge branch 'result' of https://github.com/PiotrMoscicki/PolyEngine into result

PiotrSquaresMoscicki committed:
Compilation fixes

PiotrSquaresMoscicki committed:
Fix mac compilation

PiotrSquaresMoscicki committed:
gcc compilation fix

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,285 @@
+
+
+
+
```
Remove extra blank lines

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,285 @@
+		// @returns valid value if has valid value, crashes otherwise :(
+		constexpr T& value() & { return std::get<T>(m_value); }
+		constexpr const T& value() const & { return std::get<T>(m_value); }
+		constexpr T&& value() && { return std::get<T>(std::move(m_value)); }
```
Return by value not by r-value ref (this disables rvo)

PiotrSquaresMoscicki replied:
rvo works only for variables which are created within a method and then returned

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,285 @@
+		// returns error if this stores an error, crashes otherwise :(
+		constexpr E& error() & { return std::get<E>(m_value); }
+		constexpr const E& error() const & { return std::get<E>(m_value); }
+		constexpr E&& error() && { return std::get<E>(std::move(m_value)); }
```
Return by value not by r-value ref.

PiotrSquaresMoscicki replied:
rvo works only for variables which are created within a method and then returned

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,285 @@
+		//		value
+		constexpr bool isErr() const { return !isOk(); }
+
+
```
Remove extra blank lines.

MuniuDev commented on PolyEngine/Core/Src/pe/core/utils/Result.hpp:
```diff
@@ -0,0 +1,285 @@
+		// returns error if this stores an error, crashes otherwise :(
+		constexpr E& error() & { return m_value.value(); }
+		constexpr const E& error() const & { return m_value.value(); }
+		constexpr E&& error() && { return std::move(m_value).value(); }
```
Return by value.

PiotrSquaresMoscicki replied:
rvo works only for variables which are created within a method and then returned

PiotrSquaresMoscicki committed:
CR fixes
